azure-iot-hub-devicesdk/samples/receive_device_method_async.py

```diff
@@ -0,0 +1,88 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import os
+import asyncio
+import json
+import logging
+import threading
+from six.moves import input
+from azure.iot.hub.devicesdk.aio import DeviceClient
+from azure.iot.hub.devicesdk.auth.authentication_provider_factory import from_connection_string
+
+logging.basicConfig(level=logging.ERROR)
+
+
+async def main():
+    # The connection string for a device should never be stored in code. For the sake of simplicity we're using an environment variable here.
+    conn_str = os.getenv("IOTHUB_DEVICE_CONNECTION_STRING")
+    # The "Authentication Provider" is the object in charge of creating authentication "tokens" for the device client.
+    auth_provider = from_connection_string(conn_str)
+    # For now, the SDK only supports MQTT as a protocol. the client object is used to interact with your Azure IoT hub.
+    # It needs an Authentication Provider to secure the communication with the hub, using either tokens or x509 certificates
+    device_client = DeviceClient.from_authentication_provider(auth_provider, "mqtt")
+
+    # connect the client.
+    await device_client.connect()
+
+    # define behavior for handling methods
+    async def method1_listener(device_client):
+        while True:
+            method_call = await device_client.receive_method("method1")  # Wait for method1 calls
+            payload = json.dumps({"result": True, "data": "some data"})  # set response payload
+            status = 200  # set return status code
+            print("executed method1")
+            await device_client.send_method_response(method_call, payload, status)  # send response
+
+    async def method2_listener(device_client):
+        while True:
+            method_call = await device_client.receive_method("method2")  # Wait for method2 calls
+            payload = json.dumps({"result": True, "data": 1234})  # set response payload
+            status = 200  # set return status code
+            print("executed method2")
+            await device_client.send_method_response(method_call, payload, status)  # send response
+
+    async def generic_method_listener(device_client):
+        while True:
+            method_call = await device_client.receive_method()  # Wait for unknown method calls
+            payload = json.dumps(
+                {"result": False, "data": "unknown method"}  # set response payload
+            )
+            status = 400  # set return status code
+            print("executed unknown method: " + method_call.name)
+            await device_client.send_method_response(method_call, payload, status)  # send response
+
+    # define behavior for halting the application
+    def stdin_listener():
+        while True:
+            selection = input("Press Q to exit\n")
```

Comment:
>exit [](start = 42, length = 4)

super super nitpicky: "Press Q to quit -> Quitting..." feels better than alternating between quitting or exiting. if you want to use "exit/exiting" everywhere you could also use spacebar or enter instead of q.